README.md
## SmartyStreets for PHP

[![Author](http://img.shields.io/badge/author-@mhetreramesh-blue.svg?style=flat-square)](https://twitter.com/mhetreramesh)
[![Latest Version on Packagist](https://img.shields.io/packagist/v/gliterd/smartystreets.svg?style=flat-square)](https://packagist.org/packages/gliterd/smartystreets)
[![Software License][ico-license]](LICENSE.md)
[![Build Status](https://img.shields.io/travis/gliterd/smartystreets/master.svg?style=flat-square)](https://travis-ci.org/gliterd/smartystreets)
[![Coverage Status][ico-scrutinizer]][link-scrutinizer]
[![Quality Score][ico-code-quality]][link-code-quality]
[![Total Downloads](https://img.shields.io/packagist/dt/gliterd/smartystreets.svg?style=flat-square)](https://packagist.org/packages/gliterd/smartystreets)

`smartystreets` is the SDK for working with SmartyStreets service.

## Install

Via Composer

``` bash
$ composer require gliterd/smartystreets
```


## Change log

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Testing

```bash
$ composer test
```

## How to use it?

```
<?php

require_once('vendor/autoload.php');

use \SmartyStreets\Service;

$service = new Service('your-auth-id', 'your-auth-token');

$location = $service->getInspectedLocation([
    'street'        => 'Vulkanstrasse',
    'street_number' => '106',
    'place'         => 'Zurich',
    'postal_code'   => '8010',
]);
```

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) and [CONDUCT](CONDUCT.md) for details.

## Security

If you discover any security related issues, please email mhetreramesh@gmail.com instead of using the issue tracker.

## Credits

- [All Contributors][link-contributors]

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

[ico-version]: https://img.shields.io/packagist/v/gliterd/smartystreets.svg?style=flat-square
[ico-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[ico-travis]: https://img.shields.io/travis/gliterd/smartystreets/master.svg?style=flat-square
[ico-scrutinizer]: https://img.shields.io/scrutinizer/coverage/g/gliterd/smartystreets.svg?style=flat-square
[ico-code-quality]: https://img.shields.io/scrutinizer/g/gliterd/smartystreets.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/gliterd/smartystreets.svg?style=flat-square

[link-packagist]: https://packagist.org/packages/gliterd/smartystreets
[link-travis]: https://travis-ci.org/gliterd/smartystreets
[link-scrutinizer]: https://scrutinizer-ci.com/g/gliterd/smartystreets/code-structure
[link-code-quality]: https://scrutinizer-ci.com/g/gliterd/smartystreets
[link-downloads]: https://packagist.org/packages/gliterd/smartystreets
[link-author]: https://github.com/gliterd
[link-contributors]: ../../contributors

CHANGELOG.md
# Changelog

All Notable changes to `gliterd/smartystreet` will be documented in this file.

Updates should follow the [Keep a CHANGELOG](http://keepachangelog.com/) principles.

## NEXT - YYYY-MM-DD

### Added
- Nothing

### Deprecated
- Nothing

### Fixed
- Nothing

### Removed
- Nothing

### Security
- Nothing


tests/ServiceTests.php
<?php

class ServiceTests extends PHPUnit_Framework_TestCase
{
    public function testServiceClass()
    {
        $this->assertTrue(true);
    }
}

src/Location/Lookup.php
<?php

namespace SmartyStreets\Location;

class Lookup
{
    private $street;
    private $streetNumber;
    private $place;
    private $postcode;
    private $country = 'Switzerland';
    private $geocode = 'true';

    public function __construct($values)
    {
        $this->setStreet(isset($values['street']) ? $values['street'] : '');
        $this->setStreetNumber(isset($values['street_number']) ? $values['street_number'] : '');
        $this->setPlace(isset($values['place']) ? $values['place'] : '');
        $this->setPostcode(isset($values['postcode']) ? $values['postcode'] : '');
        $this->setCountry(isset($values['country']) ? $values['country'] : 'Switzerland');
        $this->setGeocode(isset($values['geocodes']) ? $values['geocodes'] : 'true');
    }

    public function setStreet($value)
    {
        $this->street = $value;
    }

    public function getStreet()
    {
        return $this->street;
    }

    public function setStreetNumber($value)
    {
        $this->streetNumber = $value;
    }

    public function getStreetNumber()
    {
        return $this->streetNumber;
    }

    public function setPlace($value)
    {
        $this->place = $value;
    }

    public function getPlace()
    {
        return $this->place;
    }

    public function setPostcode($value)
    {
        $this->postcode = $value;
    }

    public function getPostcode()
    {
        return $this->postcode;
    }

    public function setCountry($value)
    {
        $this->country = $value;
    }

    public function getCountry()
    {
        return $this->country;
    }

    public function setGeocode($value)
    {
        $this->geocode = $value;
    }

    public function getGeocode()
    {
        return $this->geocode;
    }
}


src/Location/Location.php
<?php

namespace SmartyStreets\Location;

class Location
{
    private $street;
    private $streetNumber;
    private $locality;
    private $postcode;
    private $country = 'Switzerland';
    private $latitude;
    private $longitude;
    private $verified = false;

    public function __construct($locationArray)
    {
        $this->verified = (isset($locationArray['analysis']) && $locationArray['analysis']['verification_status'] && $locationArray['analysis']['verification_status'] == 'Verified') ? true : false;

        if (isset($locationArray['components']) && is_array($locationArray['components'])) {
            $this->street = isset($locationArray['components']['thoroughfare']) ? $locationArray['components']['thoroughfare'] : '';
            $this->streetNumber = isset($locationArray['components']['premise']) ? $locationArray['components']['premise'] : '';
            $this->locality = isset($locationArray['components']['locality']) ? $locationArray['components']['locality'] : '';
            $this->postcode = isset($locationArray['components']['postal_code']) ? $locationArray['components']['postal_code'] : '';
            $this->country = isset($locationArray['components']['country_iso_3']) ? $locationArray['components']['country_iso_3'] : '';
        }

        if (isset($locationArray['metadata']) && is_array($locationArray['metadata'])) {
            $this->latitude = isset($locationArray['metadata']['latitude']) ? $locationArray['metadata']['latitude'] : '';
            $this->longitude = isset($locationArray['metadata']['longitude']) ? $locationArray['metadata']['longitude'] : '';
        }
    }

    public function getStreet()
    {
        return $this->street;
    }

    public function getStreetNumber()
    {
        return $this->streetNumber;
    }

    public function getLocality()
    {
        return $this->locality;
    }

    public function getPostalCode()
    {
        return $this->postcode;
    }

    public function getCountry()
    {
        return $this->country;
    }

    public function getLatitude()
    {
        return $this->latitude;
    }

    public function getLongitude()
    {
        return $this->longitude;
    }

    public function getIsVerified()
    {
        return $this->verified;
    }
}


src/Service.php
<?php

namespace SmartyStreets;

use SmartyStreets\Location\Location;
use SmartyStreets\Location\Lookup;
use GuzzleHttp\Client;

class Service
{
    protected $authId;

    protected $authToken;

    private $simulation = false;

    public function __construct($authId, $authToken, $simulation = false)
    {
        $this->authId = $authId;
        $this->authToken = $authToken;
        $this->setSimulation($simulation);
    }

    public function setSimulation($value)
    {
        $this->simulation = $value;
    }

    public function getSimulation()
    {
        return $this->simulation;
    }

    public function getInspectedLocation($location)
    {
        try {
            if ($this->getSimulation()) {
                return new Location($this->getSampleLocation());
            }

            return new Location($this->getRequest(new Lookup($location)));
        } catch(\Exception $e) {
            return 'Error: ' .$e->getMessage();
        }
    }

    public function getRequest(Lookup $location)
    {
        $client = new Client([
            'base_uri' => 'https://international-street.api.smartystreets.com/',
        ]);
        $response = $client->request('GET', 'verify', [
            'query' => [
                'auth-id'     => $this->authId,
                'auth-token'  => $this->authToken,
                'address1'    => $location->getStreet().' '.$location->getStreetNumber(),
                'locality'    => $location->getPlace(),
                'postal_code' => $location->getPostcode(),
                'country'     => $location->getCountry(),
                'geocode'     => $location->getGeocode(),
            ],
        ]);

        $body = json_decode($response->getBody(), true);

        return reset($body);
    }

    protected function getSampleLocation()
    {
        return [
            'address1'   => 'Vulkanstrasse 106',
            'address2'   => '8010 Zurich',
            'components' => [

                'administrative_area'     => 'ZH',
                'sub_administrative_area' => 'Zurich',
                'country_iso_3'           => 'CHE',
                'locality'                => 'Vulkanstrasse',
                'postal_code'             => '8304',
                'postal_code_short'       => '8304',
                'premise'                 => '10',
                'premise_number'          => '10',
                'thoroughfare'            => 'Vulkanstrasse',

            ],
            'metadata' => [

                'latitude'              => 47.41603,
                'longitude'             => 8.57788,
                'geocode_precision'     => 'Premise',
                'max_geocode_precision' => 'DeliveryPoint',
                'address_format'        => 'thoroughfare premise|postal_code locality',
            ],
            'analysis' => [
                'verification_status'   => 'Verified',
                'address_precision'     => 'Premise',
                'max_address_precision' => 'Premise',
            ],
        ];
    }
}
